Structures API

Earthcomputer opened the pull request:
Replaces #687.

This PR adds an API to allow for custom structures to be registered, and adds APIs for setting basic behaviour of the structure not already easy to set with the vanilla API (like surface adjusting).
This API does *not* add structures to biomes, as it's easy to do that with the vanilla API.

This is the first API module I've added, so apologies for any mistakes with that.

Still to do:
- [x] Documentation
- [x] Test mod

Earthcomputer committed:
Initial structures API implementation

shartte commented:
This looks good to me.

Earthcomputer committed:
Improve generics + add a superflatFeature helper method

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+		return this;
+	}
+
+	public S register() {
```
buildAndRegister?

i509VCB replied:
Actually register may be fine alone

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	public S register() {
+		if (step == null) {
```
This is fine, but I'd advise `Objects.requireNonNull` or preconditions.

liach replied:
Yes, for these illegal states use objects check

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	public FabricStructureBuilder<FC, S> defaultConfig(StructureConfig config) {
+		this.defaultConfig = config;
```
Validate the parameter to make sure it is non-null (per use in build method)

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	public FabricStructureBuilder<FC, S> step(GenerationStep.Feature step) {
+		this.step = step;
```
Null check

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructureFeatureAccessor.java:
```diff
@@ -0,0 +1,41 @@
+
+	@Invoker
+	static <F extends StructureFeature<?>> F callRegister(String name, F structureFeature, GenerationStep.Feature step) {
+		throw new UnsupportedOperationException();
```
Assertion errors please. Also specify in message something like `Untransformed Accessor` or `This shouldn't happen`

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructuresImpl.java:
```diff
@@ -0,0 +1,27 @@
+import net.minecraft.world.gen.chunk.StructuresConfig;
+
+public class FabricStructuresImpl {
+	public static final Set<StructuresConfig> defaultStructuresConfigs = Collections.newSetFromMap(new WeakHashMap<>());
```
Fully capitalized name, this is a static final field.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+		this.structure = structure;
+	}
+
+	public static <FC extends FeatureConfig, S extends StructureFeature<FC>> FabricStructureBuilder<FC, S> create(Identifier id, S structure) {
```
Validate parameters as not null.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+import net.fabricmc.fabric.mixin.structure.StructureFeatureAccessor;
+import net.fabricmc.fabric.mixin.structure.StructuresConfigAccessor;
+
+public class FabricStructureBuilder<FC extends FeatureConfig, S extends StructureFeature<FC>> {
```
Javadoc please. Describe what each method in the builder does and which methods are required. I'd advise including a code example in class javadoc to show how to use this builder.

i509VCB replied:
Since constructor is private, I'd make this class final or make the constructor protected.

i509VCB replied:
Is there a vanilla structure builder? If not I may consifer omitting `Fabric` from the class name, but player and liach may have different opinions about that.

Earthcomputer replied:
I'll keep the `Fabric` prefix.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+		return this;
+	}
+
+	public FabricStructureBuilder<FC, S> defaultConfig(int spacing, int separation, int salt) {
```
I have no idea what `salt` does in this context. Please explain it briefly in javadoc

TheBrokenRail replied:
I think ``salt`` just makes sure it doesn't spawn in the same place as another structure with same ``separation`` and ``spacing``.

shartte replied:
You can probably just (@see StructureConfig) to the vanilla class. I contributed some Javadocs to Yarn that explain its fields (this method just constructs a StructureConfig and passes it to the other overload of `defaultConfig`).
See here: https://github.com/FabricMC/yarn/blob/20w30a/mappings/net/minecraft/world/gen/chunk/StructureConfig.mapping#L27

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+		return defaultConfig(new StructureConfig(spacing, separation, salt));
+	}
+
+	public FabricStructureBuilder<FC, S> superflatFeature(ConfiguredStructureFeature<FC, ? extends StructureFeature<FC>> superflatFeature) {
```
I assume this feature is used if the generator type is a superflat type. I'd specify that in javadoc so people know what it does.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/FlatChunkGeneratorConfigAccessor.java:
```diff
@@ -0,0 +1,34 @@
+public interface FlatChunkGeneratorConfigAccessor {
+	@Accessor("STRUCTURE_TO_FEATURES")
+	static Map<StructureFeature<?>, ConfiguredStructureFeature<?, ?>> getStructureToFeatures() {
+		throw new UnsupportedOperationException();
```
Assertion error please like the other accessor.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+
+		if (!id.toString().equals(structure.getName())) {
+			// mods should not be overriding getName, but if they do and it's incorrect, this gives an error
+			throw new IllegalStateException("Structure " + id + " has mismatching name " + structure.getName() + ". Structures should not override getName.");
```
Maybe put `getName` in quotes like this: `\"getName\"`

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,120 @@
+		return superflatFeature(structure.configure(config));
+	}
+
+	public FabricStructureBuilder<FC, S> adjustsSurface() {
```
I assume this is what causes the game to extend the walls of a villager house downwards to fill in air?

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructuresImpl.java:
```diff
@@ -0,0 +1,27 @@
+
+import net.minecraft.world.gen.chunk.StructuresConfig;
+
+public class FabricStructuresImpl {
```
final and a private constructor since this is an impl class and has no real need to be constructed.

liach replied:
Hmm, this `Impl` sounds like this is implementing an interface while this is not. Maybe `FabricStructureUtil`?

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/MixinChunkGeneratorType.java:
```diff
@@ -0,0 +1,34 @@
+public class MixinChunkGeneratorType {
+	@Inject(method = "method_30641", at = @At("RETURN"))
+	private static void onCreateCavesType(CallbackInfoReturnable<ChunkGeneratorType> cir) {
+		FabricStructuresImpl.defaultStructuresConfigs.add(cir.getReturnValue().getConfig());
```
I remember modmuss had to add something to biome api since nether chunk generator was initialized before mod init. Maybe validate this isn't called too early before chunk generator type is created?

Earthcomputer replied:
No, it's safe in this case. The `DEFAULT_STRUCTURES_CONFIGS` set stores the set of configs which copied from `StructureFeature.DEFAULT_STRUCTURES`. This is so that when a mod registers a structure after the `StructuresConfig` has been created, the `register` method loops over all existing configs and adds the structure to it. If the structure is registered before this `StructureConfig` is created, then since it's added to `StructureFeature.DEFAULT_STRUCTURES` the `StructureConfig` will copy the structure into its list anyway.

i509VCB commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/StructuresConfigAccessor.java:
```diff
@@ -0,0 +1,35 @@
+	@Mutable
+	@Accessor("DEFAULT_STRUCTURES")
+	static void setDefaultStructures(ImmutableMap<StructureFeature<?>, StructureConfig> defaultStructures) {
+		throw new UnsupportedOperationException();
```
Assertion error

i509VCB commented on fabric-structure-api-v1/build.gradle:
```diff
@@ -0,0 +1,6 @@
+version = getSubprojectVersion(project, "1.0.0")
+
+dependencies {
+	compile(project(path: ':fabric-api-base', configuration: 'dev'))
```
Hmm, what does this module use from base api? If it doesn't use anything from it, you can omit this dependency.

i509VCB reviewed (changes requested):
Just a few things. Also I'd advise making a small testmod (I dunno generate a 3x3 cube of diamond blocks in the world) to make api easier to test and validate builder generics are fine.

Also does this add a /locate entry?

i509VCB commented:
Also mark the module experimental since mojang has had a field day breaking world gen.

TheBrokenRail commented:
Since 1.16 ``/locate`` is automatic.

Earthcomputer commented:
Thanks for you review @i509VCB! I'll get round to addressing your comments soon. Also note that in the opening message I said I still had javadocs and a test mod to do.

i509VCB commented:
Yeah I noticed that after review.

Earthcomputer committed:
Initialize StructuresConfig class early to prevent its assertion failing

frqnny commented:
Can we shove a missing structures causing chunk issues mixin in this one? The one that if a modder removes a structure log will spam you with chunk issues and other stuff?

Earthcomputer commented:
There has been disagreements on how to implement that one. I tried implementing a version of i5's fix in LibStructure but it didn't work for me. Unless we can agree on a fix without too much bikeshedding, I'm going to leave it out of this PR and put it in a future PR.

liach reviewed (approved):
Looks great (i509VCB nitpicks a bit too much imo)

Earthcomputer committed:
Add a testmod

Earthcomputer committed:
Documentation and null assertions

Earthcomputer committed:
Apply review suggestions

shartte commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,201 @@
+ * StructureFeature structure = new MyStructure(DefaultFeatureConfig.CODEC);
+ * ConfiguredStructureFeature<DefaultFeatureConfig, ? extends StructureFeature<DefaultFeatureConfig>> configuredStructure
+ *     = structure.configure(new DefaultFeatureConfig());
+ * FabricStructureBuilder.create()
```
Unless I missed something, I think the arguments are missing here:
```suggestion
 * FabricStructureBuilder.create(new Identifier("mymod:mystructure"), structure)
```

Earthcomputer replied:
Oops

Earthcomputer committed:
Update fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java

Co-authored-by: shartte <shartte@users.noreply.github.com>

shartte commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,201 @@
+	 *
+	 * <p>This is a required option. Vanilla needs it to function.</p>
+	 *
+	 * @param spacing The minimum distance between 2 structures of this type.
+	 * @param separation The average distance between 2 structures of this type along the X and Z axes.
```
I think the description here was flipped.

```suggestion
	 * @param spacing The average distance between 2 structures of this type along the X and Z axes in chunks.
	 * @param separation The minimum distance between 2 structures of this type in chunks. Must be less than spacing.
```

shartte commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java:
```diff
@@ -0,0 +1,201 @@
+					.build());
+		}
+
+		// update existing structure configs
```
Feel free to ignore this one
```suggestion
		// update existing sets of structure configs
```

shartte commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/impl/structure/FabricStructureUtil.java:
```diff
@@ -0,0 +1,29 @@
+public final class FabricStructureUtil {
+	private FabricStructureUtil() { }
+
+	public static final Set<StructuresConfig> DEFAULT_STRUCTURES_CONFIGS = Collections.newSetFromMap(new WeakHashMap<>());
```
```suggestion
	// This tracks all StructuresConfig objects that have been created with the default set of structures
	// in order to add mod-created structures that are registered later
	public static final Set<StructuresConfig> DEFAULT_STRUCTURES_CONFIGS = Collections.newSetFromMap(new WeakHashMap<>());
```

shartte commented on fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/mixin/structure/MixinStructuresConfig.java:
```diff
@@ -0,0 +1,50 @@
+	@Final
+	private Map<StructureFeature<?>, StructureConfig> structures;
+
+	@Inject(method = "<init>(Z)V", at = @At("RETURN"))
```
```suggestion
	// This constructor of StructuresConfig initializes it with the default set of structures.
	// Since a mod can register its structures later, we need to keep track of the object created
	// here, so that we can add new structures to it later.
	@Inject(method = "<init>(Z)V", at = @At("RETURN"))
```

shartte reviewed (changes requested):
Left a few review comments. The only important one are the docs for spacing / separation. I think those are flipped.

Earthcomputer committed:
Apply review suggestions

Earthcomputer commented:
Not sure why the build failed

i509VCB commented:
Seems as is testmod failed to compile

shartte commented:
Should this be targeted at 1.16.1? I think that would fix the build.

Earthcomputer commented:
> Should this be targeted at 1.16.1? I think that would fix the build.

No. One of its mixins breaks if you go back past 20w29a. Also once #982 is merged, I'm going to update the test mod to use that.

shartte commented:
Alright. We will also need to add every registered structure to Biome#field_26634 then, since it copies the contents of Registry.STRUCTURE_FEATURE at the time the Biome is constructed.
(p.s.: This field contains _all_ structures, not just the ones that can start in the Biome)

i509VCB commented on fabric-structure-api-v1/src/testmod/java/net/fabricmc/fabric/test/structure/mixin/MixinDefaultBiomeCreator.java:
```diff
@@ -0,0 +1,37 @@
+	@Inject(method = "createPlains", at = @At("RETURN"))
+	private static void addCustomStructure(String parent, boolean sunflower, CallbackInfoReturnable<Biome> cir) {
+		if (!sunflower) {
+			cir.getReturnValue().addStructureFeature(StructureTest.CONFIGURED_STRUCTURE);
```
Hopefully modmuss's biome pr will remove the need for this.

Player3324 reviewed (approved):
LGTM, thanks! I think we need a copy of this PR for both 1.16.2 and 1.15.2 to make the merge easier on us.

modmuss50 reviewed (approved):
Looks good, just needs testing outside of dev (I can do this before merging).

And yes, like player said a 1.15.2 PR would be greatly appreicated if its not too much work. (We can backport that to 1.14 if still applicable)

Earthcomputer commented:
Since the 1.16.2-pre1 update, this also needs a fix to add structures to the existing lists copied in biomes. I can add this real quick, see the discussion in discord.

Earthcomputer committed:
Merge branch '1.16' into feature/structures

Earthcomputer committed:
Update to 1.16.2

Earthcomputer commented:
There we go.

Earthcomputer commented:
Re: copies of this PR for 1.15 and 1.14, I don't personally think this is necessary, as this PR was made to get around the fact that registering structures is very difficult in 1.16+. In 1.15 and below doing it using the vanilla API is not hard at all, and this PR doesn't make it easier and not all builder methods are even 100% applicable.

modmuss50 commented:
Ok, cool. This will be for 1.16 only then.
